[SQL] REPEAT(binary) fails with compilation error

Following SQL:

```
CREATE TABLE tbl(bin BINARY);

CREATE MATERIALIZED VIEW v AS SELECT
REPEAT(bin, 2) AS bin
FROM tbl;
```

Given example input:
**"bin": [11, 22, 32]**

Fails with the following compilation error:
```
error[E0308]: mismatched types
   --> crates/feldera_pipe_operator_d70a02fbee938765/src/lib.rs:91:17
    |
90  |             repeatN_(
    |             -------- arguments to this function are incorrect
91  |                 (*p0).6.as_ref().cloned(),
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Option<SqlString>`, found `Option<ByteArray>`
    |
    = note: expected enum `std::option::Option<feldera_sqllib::SqlString>`
               found enum `std::option::Option<feldera_sqllib::ByteArray>`
note: function defined here
   --> /Users/rivuadhikari/feldera/crates/sqllib/src/string.rs:299:1
    |
299 | some_function2!(repeat, SqlString, i32, SqlString);
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: this error originates in the macro `some_function2` (in Nightly builds, run with -Z macro-backtrace for more info)
```

--- a/crates/sqllib/src/string.rs
+++ b/crates/sqllib/src/string.rs
@@ -160,27 +160,21 @@ some_function2!(substring2, SqlString, i32, SqlString);
 
 #[doc(hidden)]
 pub fn trim_both_s_s(remove: SqlString, value: SqlString) -> SqlString {
-    // 'remove' always has exactly 1 character
-    let chr = remove.str().chars().next().unwrap();
-    SqlString::from(value.str().trim_matches(chr))
+    SqlString::from(value.str().trim_matches(|c| remove.str().contains(c)))
 }
 
 some_polymorphic_function2!(trim_both, s, SqlString, s, SqlString, SqlString);
 
 #[doc(hidden)]
 pub fn trim_leading_s_s(remove: SqlString, value: SqlString) -> SqlString {
-    // 'remove' always has exactly 1 character
-    let chr = remove.str().chars().next().unwrap();
-    SqlString::from(value.str().trim_start_matches(chr))
+    SqlString::from(value.str().trim_start_matches(|c| remove.str().contains(c)))
 }
 
 some_polymorphic_function2!(trim_leading, s, SqlString, s, SqlString, SqlString);
 
 #[doc(hidden)]
 pub fn trim_trailing_s_s(remove: SqlString, value: SqlString) -> SqlString {
-    // 'remove' always has exactly 1 character
-    let chr = remove.str().chars().next().unwrap();
-    SqlString::from(value.str().trim_end_matches(chr))
+    SqlString::from(value.str().trim_end_matches(|c| remove.str().contains(c)))
 }
 
 some_polymorphic_function2!(trim_trailing, s, SqlString, s, SqlString, SqlString);

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/ExpressionCompiler.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/ExpressionCompiler.java
@@ -1155,9 +1155,16 @@ public class ExpressionCompiler extends RexVisitorImpl<DBSPExpression>
                         return compileFunction(call, node, type, ops, 3);
                     }
                     case "overlay": {
+                        validateArgCount(node, opName, ops.size(), 3, 4);
                         String module_prefix = "";
                         if (ops.get(0).type.is(DBSPTypeBinary.class)) {
                             module_prefix = "binary_";
+                        }
+                        if (ops.get(0).getType().code != ops.get(1).getType().code) {
+                            throw new CompilationError(
+                                    "First and second arguments to 'OVERLAY' must have the same type; the types are\n" +
+                                    ops.get(0).getType().asSqlString() + " and " + ops.get(1).getType().asSqlString(),
+                                    node);
                         }
                         this.ensureInteger(node, ops, 2);
                         if (ops.size() == 4)
@@ -1201,6 +1208,10 @@ public class ExpressionCompiler extends RexVisitorImpl<DBSPExpression>
                     case "writelog":
                         return new DBSPApplyExpression(node, opName, type, ops.get(0), ops.get(1));
                     case "repeat": {
+                        validateArgCount(node, opName, ops.size(), 2);
+                        if (ops.get(0).getType().code == BYTES) {
+                            throw new UnimplementedException("'REPEAT' with a VAR/BINARY argument not yet supported", node);
+                        }
                         this.ensureInteger(node, ops, 1);
                         return compileFunction(call, node, type, ops, 2);
                     }

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/Regression1Tests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/Regression1Tests.java
@@ -425,4 +425,44 @@ public class Regression1Tests extends SqlIoTest {
                 CREATE TABLE illegal_tbl(bin BINARY);
                 CREATE VIEW v AS SELECT CONCAT_WS('@', bin, 55) AS bin FROM illegal_tbl;""");
     }
+
+    @Test
+    public void issue4265() {
+        this.statementsFailingInCompilation("""
+                CREATE TABLE tbl(bin BINARY);
+                CREATE VIEW v AS SELECT
+                OVERLAY(bin placing 'i' from 2 for 4) AS bin FROM tbl;""",
+                "First and second arguments to 'OVERLAY' must have the same type");
+    }
+
+    @Test
+    public void issue4263() {
+        this.statementsFailingInCompilation("""
+                CREATE TABLE tbl(bin BINARY);
+                CREATE VIEW v AS SELECT
+                REPEAT(bin, 2) AS bin FROM tbl;
+                """, "Not yet implemented: 'REPEAT' with a VAR/BINARY argument not yet supported");
+    }
+
+    @Test
+    public void issue4264() {
+        this.qs("""
+                SELECT TRIM(trailing '' FROM 'x');
+                 result
+                --------
+                 x
+                (1 row)
+                
+                SELECT TRIM(leading 'abc' FROM 'abacabadaba');
+                 result
+                --------
+                 daba
+                (1 row)
+                
+                SELECT TRIM(both 'abc' FROM 'abacabadaba');
+                 result
+                --------
+                 d
+                (1 row)""");
+    }
 }

